docs/source/tutorials/qmd/07-error-correction.qmd

```diff
@@ -0,0 +1,623 @@
+
+$$
+\tilde{x}_i^{(m)} = x_i + u_i^{(m)}, \quad m = 1, \ldots, M.
+$$
+
+The measurement error is assumed to follow a multivariate Gaussian model,
+such that
+
+$$
+\mathbf{u}_i \sim N_M(\mathbf{0}, \mathbf{\Sigma}_{u,i}).
+$$
+
+The fundamental concept behind Bayesian measurement error correction is
+to treat the true, unknown covariate values $x_i$ as additional latent variables.
+These values are then imputed using MCMC simulations while simultaneously
+estimating all other parameters of the model. To achieve this, a prior distribution
+is assigned to $x_i$. Additional flexibility is achieved by adding hyperpriors with
+assigned prior distributions to the distribution parameters of $x_i$.
+$$
+x_i \sim N(\mu_x,\tau^2_x), \quad \mu_x \sim N(0, \tau^2_{\mu}), \quad \tau^2_x \sim IG(a_x,b_x)
+$$
+
+
+```{python}
+#| label: imports
+import jax
+import jax.numpy as jnp
+import liesel.model as lsl
+import liesel.goose as gs
+import liesel.contrib.splines as splines
+import tensorflow_probability.substrates.jax as tfp
+import tensorflow_probability.substrates.jax.bijectors as tfb
+import numpy as np
+import matplotlib.pyplot as plt
+
+from liesel.contrib.splines import equidistant_knots, basis_matrix
+from liesel.distributions.mvn_degen import MultivariateNormalDegenerate
+
+tfd = tfp.distributions
+```
+## Data
+
+We will start by simulating some data with replicates. To introduce heteroscedasticity we scale the first half of the
+covariance matrices by one and the second by two.
+
+```{python}
+#| label: data
+# Define the number of samples and replicates
+n = 500  # Number of data points
+M = 3    # Number of replicates per sample
+c_u = 0.75 # define measurement error covariance
+
+np.random.seed(123)
+
+x = np.random.normal(loc = 10, scale = 5, size = n)
+
+# Generate covariance matrices
+def create_sigma_me(dim_sigma_me, c_u):
+    Sigma_me = np.zeros((dim_sigma_me, dim_sigma_me))
+    for i in range(dim_sigma_me):
+        for j in range(dim_sigma_me):
+            if i == j:
+                Sigma_me[i, j] = 1
+            else:
+                Sigma_me[i, j] = c_u
+    return Sigma_me
+
+# Scale the Sigma_me matrices by individual variance factor
+sigma_matrices = []
+for i in range(n):
+    sigma_sq_ui = 1 if i < n // 2 else 2  # First half scaled by 1, second half by 2
+    Sigma_me = create_sigma_me(M, c_u)
+    scaled_matrix = sigma_sq_ui * Sigma_me
+    sigma_matrices.append(scaled_matrix)
+
+    sigma_matrices_array = jnp.array(np.stack(sigma_matrices))
+
+# Create M replicates of true variable x
+replicates = []
+for i in range(n):
+    mean_vector = np.repeat(x[i], M)
+    Sigma_me = sigma_matrices[i]
+    samples = np.random.multivariate_normal(mean_vector, Sigma_me)
+    replicates.append(samples)
+
+# Generate response variable y
+variances = np.random.choice([0.3, 0.5], size = n)
+y_true = np.sin(x)
+y = np.random.normal(loc = np.sin(x), scale = np.sqrt(variances))
+
+# Convert to JAX arrays
```

Comment:
Man könnte auch die ganze Erzeugung in Jax durchführen, dann wäre keine Konvertierung hier mehr notwendig. Das fänd ich ganz nett :)